File: recipes/subms-segment-reader/java/src/main/java/com/submillisecond/recipes/segment/SampleApp.java
package com.submillisecond.recipes.segment;

import com.submillisecond.recipes.segment.features.Crc32SegmentReader;
import com.submillisecond.recipes.segment.features.Crc32SegmentWriter;
import com.submillisecond.recipes.segment.features.IndexedSegmentReader;
import com.submillisecond.recipes.segment.features.Lz4BlockWriter;
import com.submillisecond.recipes.segment.features.Lz4SegmentReader;
import com.submillisecond.recipes.segment.features.MmapSegmentReader;
import com.submillisecond.recipes.segment.features.WalCursorReader;
import com.submillisecond.recipes.segment.features.Xxh3SegmentReader;
import com.submillisecond.recipes.segment.features.Xxh3SegmentWriter;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Sample app: a tour of {@code subms-segment-reader} against an order-event
 * journal, base API first, then each optional reader. Run:
 * {@code mvn -q compile exec:java -Dexec.mainClass=com.submillisecond.recipes.segment.SampleApp}
 *
 * <ul>
 *   <li>base        - replay an order journal, stopping cleanly on a crash-torn tail
 *   <li>mmap        - map a large capture file and page it in lazily
 *   <li>crc32       - detect a bit-flip in an at-rest capture (RocksDB-style trailer)
 *   <li>xxh3        - faster integrity check for a trusted internal pipeline
 *   <li>lz4         - compressed capture blocks that decompress on read
 *   <li>seek-index  - jump to the millionth event without scanning from the head
 *   <li>wal-cursor  - replay only what the writer has fsync'd, never a partial tail
 * </ul>
 */
public final class SampleApp {

    public static void main(String[] args) throws IOException {
        baseJournalReplay();
        mmapCaptureReplay();
        crc32CorruptionGuard();
        xxh3FastIntegrity();
        lz4CompressedBlocks();
        seekToEvent();
        durableReplay();
    }

    /** Encode a run of order events into a framed segment. */
    static byte[] buildJournal(String... events) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        SegmentWriter w = new SegmentWriter(dos);
        for (String e : events) w.write(e.getBytes(StandardCharsets.UTF_8));
        dos.close();
        return baos.toByteArray();
    }

    /** Base API: a matching engine replays its order journal on restart.
     *  The process died mid-append last time, so the segment ends torn. */
    static void baseJournalReplay() throws IOException {
        System.out.println("== base: order-journal replay after a crash ==");
        byte[] intact = buildJournal(
            "NEW  AAPL  buy  100 @ 150.25",
            "NEW  MSFT  sell  50 @ 402.10",
            "CXL  AAPL  ord-1");
        // The writer was cut mid-frame: a header claiming 32 bytes, 4 written.
        byte[] segment = new byte[intact.length + 8];
        System.arraycopy(intact, 0, segment, 0, intact.length);
        segment[intact.length + 3] = 32;
        segment[intact.length + 4] = 'N';
        segment[intact.length + 5] = 'E';
        segment[intact.length + 6] = 'W';
        segment[intact.length + 7] = ' ';

        SegmentReader r = new SegmentReader(new DataInputStream(new ByteArrayInputStream(segment)));
        int replayed = 0;
        while (true) {
            try {
                byte[] event = r.nextRecord();
                if (event == null) {
                    System.out.println("  -> clean EOF");
                    break;
                }
                System.out.println("  replay " + new String(event, StandardCharsets.UTF_8));
                replayed++;
            } catch (SegmentReader.TruncatedFrame torn) {
                System.out.println("  -> stopped at torn tail: " + torn.getMessage());
                break;
            }
        }
        System.out.println("  replayed " + replayed + " intact events");
        if (replayed != 3) throw new AssertionError("every event before the torn tail is recovered");
    }

    /** mmap: map a capture file in constant time; the OS pages in only the
     *  frames actually touched. */
    static void mmapCaptureReplay() throws IOException {
        System.out.println("\n== mmap: replay a capture file without loading it whole ==");
        byte[] segment = buildJournal("TICK AAPL 150.25", "TICK AAPL 150.26", "TICK MSFT 402.11");
        Path path = Files.createTempFile("subms-segment-sample-", ".capture");
        Files.write(path, segment);

        int ticks = 0;
        try (MmapSegmentReader r = new MmapSegmentReader(path)) {
            System.out.println("  mapped " + r.length() + " bytes, resident set grows on touch");
            byte[] event;
            while ((event = r.nextRecord()) != null) {
                System.out.println("  " + new String(event, StandardCharsets.UTF_8));
                ticks++;
            }
        } finally {
            Files.deleteIfExists(path);
        }
        if (ticks != 3) throw new AssertionError("mmap path reads the same frames as the base reader");
    }

    /** crc32: a CRC32C trailer per block turns silent corruption into a
     *  typed ChecksumMismatch on read. */
    static void crc32CorruptionGuard() throws IOException {
        System.out.println("\n== crc32: catch a bit-flip in an at-rest capture ==");
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        Crc32SegmentWriter w = new Crc32SegmentWriter(dos);
        w.write("FILL AAPL 100 @ 150.25".getBytes(StandardCharsets.UTF_8));
        w.write("FILL MSFT  50 @ 402.10".getBytes(StandardCharsets.UTF_8));
        dos.close();
        byte[] segment = baos.toByteArray();
        segment[4] ^= 0x08; // flip a bit in the first payload

        Crc32SegmentReader r = new Crc32SegmentReader(
            new DataInputStream(new ByteArrayInputStream(segment)));
        try {
            r.nextRecord();
            throw new AssertionError("expected ChecksumMismatch on corrupted block");
        } catch (Crc32SegmentReader.ChecksumMismatch expected) {
            System.out.println("  block 0: ChecksumMismatch (corruption caught)");
        }
    }

    /** xxh3: a cheaper integrity check for a trusted, single-implementation
     *  pipeline. Catches accidental corruption; not adversary-safe. */
    static void xxh3FastIntegrity() throws IOException {
        System.out.println("\n== xxh3: faster per-block integrity on a trusted pipeline ==");
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        Xxh3SegmentWriter w = new Xxh3SegmentWriter(dos);
        for (String e : new String[] {"FILL AAPL 100 @ 150.25", "FILL AAPL  25 @ 150.26"}) {
            w.write(e.getBytes(StandardCharsets.UTF_8));
        }
        dos.close();

        Xxh3SegmentReader r = new Xxh3SegmentReader(
            new DataInputStream(new ByteArrayInputStream(baos.toByteArray())));
        int ok = 0;
        byte[] event;
        while ((event = r.nextRecord()) != null) {
            System.out.println("  verified " + new String(event, StandardCharsets.UTF_8));
            ok++;
        }
        if (ok != 2) throw new AssertionError("both blocks pass their xxh3 trailer");
    }

    /** lz4: repetitive order events compress well; the reader decompresses
     *  transparently on read. */
    static void lz4CompressedBlocks() throws IOException {
        System.out.println("\n== lz4: compressed capture blocks ==");
        byte[] payload = "NEW AAPL buy 100 @ 150.25\n".repeat(256).getBytes(StandardCharsets.UTF_8);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        new Lz4BlockWriter(dos).write(payload);
        dos.close();
        byte[] segment = baos.toByteArray();
        System.out.println("  " + payload.length + " raw bytes stored in " + segment.length + " on-disk bytes");
        if (segment.length >= payload.length) throw new AssertionError("repetitive block should compress");

        Lz4SegmentReader r = new Lz4SegmentReader(
            new DataInputStream(new ByteArrayInputStream(segment)));
        byte[] block = r.nextRecord();
        if (!java.util.Arrays.equals(block, payload)) {
            throw new AssertionError("decompresses byte-for-byte");
        }
        System.out.println("  decompressed back to " + block.length + " bytes");
    }

    /** seek-index: index every 64th block at open, then binary-search plus a
     *  bounded scan to reach any block. */
    static void seekToEvent() throws IOException {
        System.out.println("\n== seek-index: jump to event 100 without scanning the head ==");
        String[] events = new String[200];
        for (int i = 0; i < events.length; i++) events[i] = "SEQ " + i + ": TICK AAPL";
        byte[] segment = buildJournal(events);

        IndexedSegmentReader r = new IndexedSegmentReader(segment);
        System.out.println("  " + r.totalBlocks() + " blocks, " + r.indexLen() + " sparse index entries");
        r.seekToBlock(100);
        byte[] at = r.nextRecord();
        System.out.println("  seek(100) -> " + new String(at, StandardCharsets.UTF_8));
        if (!java.util.Arrays.equals(at, "SEQ 100: TICK AAPL".getBytes(StandardCharsets.UTF_8))) {
            throw new AssertionError("landed on the requested event");
        }
    }

    /** wal-cursor: readCommitted() stops at the durable watermark; advancing
     *  it exposes the newly-committed prefix. */
    static void durableReplay() throws IOException {
        System.out.println("\n== wal-cursor: replay only the fsync'd prefix ==");
        byte[] segment = buildJournal("FILL ord-1", "FILL ord-2", "FILL ord-3");
        int afterFirst = 4 + "FILL ord-1".length();

        WalCursorReader r = new WalCursorReader(segment);
        if (r.readCommitted() != null) throw new AssertionError("watermark at 0 -> nothing durable");

        r.setCommitted(afterFirst);
        byte[] event = r.readCommitted();
        System.out.println("  after first fsync -> " + new String(event, StandardCharsets.UTF_8));
        if (r.readCommitted() != null) throw new AssertionError("second block not yet committed");

        r.setCommitted(segment.length);
        int tail = 0;
        while ((event = r.readCommitted()) != null) {
            System.out.println("  now durable -> " + new String(event, StandardCharsets.UTF_8));
            tail++;
        }
        if (tail != 2) throw new AssertionError("the remaining committed events replay");
    }

    private SampleApp() {}
}

File: recipes/subms-segment-reader/rust/examples/sample_app.rs
//! Sample app: a tour of `subms-segment-reader` against an order-event
//! journal, base API first, then each optional reader. Run the base with
//! `cargo run --example sample_app`; add `--all-features` (or a subset like
//! `--features crc32,seek-index`) to light up the feature sections.
//!
//! * base        - replay an order journal, stopping cleanly on a crash-torn tail
//! * mmap        - map a large capture file and page it in lazily
//! * crc32       - detect a bit-flip in an at-rest capture (RocksDB-style trailer)
//! * xxh3        - faster integrity check for a trusted internal pipeline
//! * lz4         - compressed capture blocks that decompress on read
//! * seek-index  - jump to the millionth event without scanning from the head
//! * wal-cursor  - replay only what the writer has fsync'd, never a partial tail

use subms_segment_reader::{SegmentReader, SegmentWriter};

fn main() {
    base_journal_replay();

    #[cfg(feature = "mmap")]
    mmap_capture_replay();

    #[cfg(feature = "crc32")]
    crc32_corruption_guard();

    #[cfg(feature = "xxh3")]
    xxh3_fast_integrity();

    #[cfg(feature = "lz4")]
    lz4_compressed_blocks();

    #[cfg(feature = "seek-index")]
    seek_to_event();

    #[cfg(feature = "wal-cursor")]
    durable_replay();
}

/// Encode a run of order events into a framed segment. Each record is one
/// line of an order journal; the writer just length-prefixes them.
fn build_journal(events: &[&str]) -> Vec<u8> {
    let mut buf = Vec::new();
    let mut w = SegmentWriter::new(&mut buf);
    for e in events {
        w.write(e.as_bytes()).unwrap();
    }
    buf
}

/// Base API: a matching engine replays its order journal on restart. The
/// process died mid-append last time, so the segment ends in a torn frame.
/// Replay walks every intact event and stops cleanly on the torn tail
/// instead of panicking.
fn base_journal_replay() {
    println!("== base: order-journal replay after a crash ==");
    let mut segment = build_journal(&[
        "NEW  AAPL  buy  100 @ 150.25",
        "NEW  MSFT  sell  50 @ 402.10",
        "CXL  AAPL  ord-1",
    ]);
    // The writer was cut mid-frame: a length header claiming 32 bytes, 4 written.
    segment.extend_from_slice(&[0, 0, 0, 32]);
    segment.extend_from_slice(b"NEW ");

    let mut r = SegmentReader::new(segment.as_slice());
    let mut replayed = 0usize;
    loop {
        match r.next_record() {
            Ok(Some(event)) => {
                println!("  replay {}", String::from_utf8_lossy(event));
                replayed += 1;
            }
            Ok(None) => {
                println!("  -> clean EOF");
                break;
            }
            Err(e) => {
                println!("  -> stopped at torn tail: {e}");
                break;
            }
        }
    }
    println!("  replayed {replayed} intact events");
    assert_eq!(replayed, 3, "every event before the torn tail is recovered");
}

/// `mmap` feature: a day's market-data capture is far larger than the
/// working set of any one replay. `MmapSegmentReader::open` maps it in
/// constant time and the OS pages in only the frames actually touched.
#[cfg(feature = "mmap")]
fn mmap_capture_replay() {
    use std::io::Write;
    use subms_segment_reader::MmapSegmentReader;
    println!("\n== mmap: replay a capture file without loading it whole ==");

    let segment = build_journal(&["TICK AAPL 150.25", "TICK AAPL 150.26", "TICK MSFT 402.11"]);
    let mut path = std::env::temp_dir();
    path.push(format!(
        "subms-segment-sample-{}.capture",
        std::process::id()
    ));
    std::fs::File::create(&path)
        .unwrap()
        .write_all(&segment)
        .unwrap();

    let mut r = MmapSegmentReader::open(&path).unwrap();
    println!("  mapped {} bytes, resident set grows on touch", r.len());
    let mut ticks = 0usize;
    while let Some(event) = r.next_record().unwrap() {
        println!("  {}", String::from_utf8_lossy(event));
        ticks += 1;
    }
    std::fs::remove_file(&path).ok();
    assert_eq!(
        ticks, 3,
        "mmap path reads the same frames as the base reader"
    );
}

/// `crc32` feature: an at-rest capture can bit-rot on disk. A CRC32C trailer
/// per block (the LevelDB / RocksDB / Parquet standard) turns silent
/// corruption into a typed `ChecksumMismatch` on read.
#[cfg(feature = "crc32")]
fn crc32_corruption_guard() {
    use subms_segment_reader::Crc32SegmentReader;
    use subms_segment_reader::Error;
    use subms_segment_reader::features::crc32::Crc32SegmentWriter;
    println!("\n== crc32: catch a bit-flip in an at-rest capture ==");

    let mut segment = Vec::new();
    {
        let mut w = Crc32SegmentWriter::new(&mut segment);
        w.write(b"FILL AAPL 100 @ 150.25").unwrap();
        w.write(b"FILL MSFT  50 @ 402.10").unwrap();
    }
    // Flip one bit in the first payload (a cosmic-ray / bad-sector stand-in).
    segment[4] ^= 0x08;

    let mut r = Crc32SegmentReader::new(segment.as_slice());
    match r.next_record() {
        Err(Error::ChecksumMismatch) => println!("  block 0: ChecksumMismatch (corruption caught)"),
        other => panic!("expected ChecksumMismatch, got {other:?}"),
    }
}

/// `xxh3` feature: inside a trusted, single-implementation pipeline where
/// throughput beats interoperability, xxHash3-64 is the cheaper integrity
/// check. It still catches accidental corruption; it is not adversary-safe.
#[cfg(feature = "xxh3")]
fn xxh3_fast_integrity() {
    use subms_segment_reader::Xxh3SegmentReader;
    use subms_segment_reader::features::xxh3::Xxh3SegmentWriter;
    println!("\n== xxh3: faster per-block integrity on a trusted pipeline ==");

    let mut segment = Vec::new();
    {
        let mut w = Xxh3SegmentWriter::new(&mut segment);
        for e in ["FILL AAPL 100 @ 150.25", "FILL AAPL  25 @ 150.26"] {
            w.write(e.as_bytes()).unwrap();
        }
    }
    let mut r = Xxh3SegmentReader::new(segment.as_slice());
    let mut ok = 0usize;
    while let Some(event) = r.next_record().unwrap() {
        println!("  verified {}", String::from_utf8_lossy(event));
        ok += 1;
    }
    assert_eq!(ok, 2, "both blocks pass their xxh3 trailer");
}

/// `lz4` feature: order events are highly repetitive, so the capture
/// compresses well. `Lz4BlockWriter` picks the smaller of stored / lz4 per
/// block; `Lz4SegmentReader` decompresses transparently on read.
#[cfg(feature = "lz4")]
fn lz4_compressed_blocks() {
    use subms_segment_reader::{Lz4BlockWriter, Lz4SegmentReader};
    println!("\n== lz4: compressed capture blocks ==");

    let payload = "NEW AAPL buy 100 @ 150.25\n".repeat(256);
    let mut segment = Vec::new();
    Lz4BlockWriter::new(&mut segment)
        .write(payload.as_bytes())
        .unwrap();
    println!(
        "  {} raw bytes stored in {} on-disk bytes",
        payload.len(),
        segment.len()
    );
    assert!(segment.len() < payload.len(), "repetitive block compressed");

    let mut r = Lz4SegmentReader::new(segment.as_slice());
    let block = r.next_record().unwrap().unwrap();
    assert_eq!(block, payload.as_bytes(), "decompresses byte-for-byte");
    println!("  decompressed back to {} bytes", block.len());
}

/// `seek-index` feature: a replay tool wants event 100 out of a long
/// capture without scanning the first 99. `IndexedSegmentReader` indexes
/// every 64th block at open, then binary-searches + scans a bounded tail.
#[cfg(feature = "seek-index")]
fn seek_to_event() {
    use subms_segment_reader::IndexedSegmentReader;
    println!("\n== seek-index: jump to event 100 without scanning the head ==");

    let events: Vec<String> = (0..200).map(|i| format!("SEQ {i}: TICK AAPL")).collect();
    let refs: Vec<&str> = events.iter().map(String::as_str).collect();
    let segment = build_journal(&refs);

    let mut r = IndexedSegmentReader::open(&segment).unwrap();
    println!(
        "  {} blocks, {} sparse index entries",
        r.total_blocks(),
        r.index_len()
    );
    r.seek_to_block(100).unwrap();
    let at = r.next_record().unwrap().unwrap();
    println!("  seek(100) -> {}", String::from_utf8_lossy(at));
    assert_eq!(at, b"SEQ 100: TICK AAPL", "landed on the requested event");
}

/// `wal-cursor` feature: a live tailer must never replay an event the
/// writer has not yet fsync'd. `read_committed()` stops at the durable
/// watermark; advancing it exposes the newly-committed prefix.
#[cfg(feature = "wal-cursor")]
fn durable_replay() {
    use subms_segment_reader::WalCursorReader;
    println!("\n== wal-cursor: replay only the fsync'd prefix ==");

    let segment = build_journal(&["FILL ord-1", "FILL ord-2", "FILL ord-3"]);
    // Byte offset of the tail of the first block: 4-byte header + 10-byte payload.
    let after_first = 4 + "FILL ord-1".len();

    let mut r = WalCursorReader::new(&segment);
    assert!(
        r.read_committed().unwrap().is_none(),
        "watermark at 0 -> nothing durable"
    );

    r.set_committed(after_first);
    let event = r.read_committed().unwrap().unwrap();
    println!("  after first fsync -> {}", String::from_utf8_lossy(event));
    assert!(
        r.read_committed().unwrap().is_none(),
        "second block not yet committed"
    );

    r.set_committed(segment.len());
    let mut tail = 0usize;
    while let Some(event) = r.read_committed().unwrap() {
        println!("  now durable -> {}", String::from_utf8_lossy(event));
        tail += 1;
    }
    assert_eq!(tail, 2, "the remaining committed events replay");
}
